pkg/networkservice/connectioncontextkernel/ethernetcontext/setmacaddress/server.go

```diff
@@ -57,26 +60,32 @@ type setVppMacServer struct{}
 //                              +---------------------------+
 //
 func NewServer() networkservice.NetworkServiceServer {
-	return &setVppMacServer{}
+	return &setKernelMacServer{}
 }
 
-func (s *setVppMacServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*connection.Connection, error) {
-	conf := vppagent.Config(ctx)
-	conn, err := next.Client(ctx).Request(ctx, request)
-	if err != nil {
-		return nil, err
-	}
+func (s *setKernelMacServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*connection.Connection, error) {
 	if mechanism := kernel.ToMechanism(request.GetConnection().GetMechanism()); mechanism != nil {
-		index := len(conf.GetLinuxConfig().GetInterfaces()) - 1
-		conf.GetLinuxConfig().GetInterfaces()[index+1].PhysAddress = conn.GetContext().GetEthernetContext().GetDstMac()
+		s.setDstMac(ctx, request.GetConnection())
+	}
+	return next.Server(ctx).Request(ctx, request)
+}
+
+func (s *setKernelMacServer) Close(ctx context.Context, conn *connection.Connection) (*empty.Empty, error) {
+	if mechanism := kernel.ToMechanism(conn.GetMechanism()); mechanism != nil {
+		s.setDstMac(ctx, conn)
 	}
-	return conn, nil
+	return next.Server(ctx).Close(ctx, conn)
 }
 
-func (s *setVppMacServer) Close(ctx context.Context, conn *connection.Connection) (*empty.Empty, error) {
-	conf := vppagent.Config(ctx)
-	if mechanism := kernel.ToMechanism(conn.GetMechanism()); mechanism != nil && len(conf.GetLinuxConfig().GetInterfaces()) > 0 {
-		conf.GetLinuxConfig().GetInterfaces()[len(conf.GetVppConfig().GetInterfaces())-1].PhysAddress = conn.GetContext().GetEthernetContext().GetDstMac()
+func (s *setKernelMacServer) setDstMac(ctx context.Context, conn *connection.Connection) {
+	ethernetContext := conn.GetContext().GetEthernetContext()
+	config := vppagent.Config(ctx)
+	if ethernetContext != nil && ethernetContext.DstMac != "" {
+		interfaceName := fmt.Sprintf("DST-%v", conn.GetId())
+		for _, iface := range config.LinuxConfig.Interfaces {
```

Comment:
Why are we looping over all interfaces here?

Reply:
To exclude setting mac address for veth interfaces for example

Reply:
The problem is it relies on a naming convention, one that is already different from what is here:

https://github.com/edwarnicke/sdk-vppagent/blob/master/pkg/networkservice/mechanisms/kernel/kernelvethpair/server.go#L45

Rather than baking naming conventions in, the previous approach was simply this: the last Linux Interface is the one we apply config to.

Reply:
Do you suggest use index convention?

Reply:
Using the last interface looking is not clear to me. Is it some pattern?
Can we just put the current interface into context? For example

context.WithDestenationInterface(ctx, inteface)
...
interface := vppagent.DestenationInterface(ctx)
interface.Set(...)

Reply:
Finally got it